Search block: Add typography supports

ramonjd opened the pull request:
Related:
- https://github.com/WordPress/gutenberg/issues/43241
- https://github.com/WordPress/gutenberg/issues/43242

## What?

Adds typography supports to the Search block.

## Why?

- Allows typographic styling of the Search block.
- Improves consistency of our design tools across blocks.

## How?

- Opts into typography supports and applies them to the: 
    - label
    - input (except text decoration)
    - button
- Sets the font size control to display by default matching most other blocks.

## Testing Instructions

1. Edit a post, add a Search block, and select it. (See example HTML below).
2. Check that all typography controls are now available.
3. Test various typography settings ensuring styles are applied in the editor.
4. Test various search block manifestations, e.g., button position, icon button.
5. Save and confirm that the frontend appears as expected
6. Do the same in the site editor. Also, edit global styles and make sure the styles are applied.
7. Style the Search block via theme.json (see example json below).
8. Confirm the theme.json style applies correctly in the editor and frontend.

<details>

<summary>Example block code</summary>

```html
<!-- wp:search {"label":"Search","buttonText":"Search","style":{"typography":{"fontStyle":"italic","fontWeight":"600","textTransform":"uppercase","letterSpacing":"55px","lineHeight":3.5,"fontSize":"55px","textDecoration":"underline"}},"fontFamily":"source-serif-pro"} /-->

<!-- wp:separator {"className":"is-style-wide"} -->
<hr class="wp-block-separator has-alpha-channel-opacity is-style-wide"/>
<!-- /wp:separator -->

<!-- wp:search {"label":"Search","showLabel":false,"buttonText":"Search","buttonPosition":"button-inside","style":{"typography":{"fontStyle":"italic","fontWeight":"600","textTransform":"capitalize","letterSpacing":"55px","lineHeight":3.5,"fontSize":"55px","textDecoration":"line-through"}},"fontFamily":"source-serif-pro"} /-->

<!-- wp:separator {"className":"is-style-wide"} -->
<hr class="wp-block-separator has-alpha-channel-opacity is-style-wide"/>
<!-- /wp:separator -->

<!-- wp:search {"label":"Search","buttonText":"Search","style":{"typography":{"fontStyle":"italic","fontWeight":"600","textTransform":"capitalize","letterSpacing":"55px","lineHeight":3.5,"fontSize":"55px","textDecoration":"line-through"}},"fontFamily":"source-serif-pro"} /-->

<!-- wp:separator {"className":"is-style-wide"} -->
<hr class="wp-block-separator has-alpha-channel-opacity is-style-wide"/>
<!-- /wp:separator -->

<!-- wp:search {"label":"Search","buttonText":"Search","buttonUseIcon":true,"style":{"typography":{"fontStyle":"italic","fontWeight":"600","textTransform":"capitalize","letterSpacing":"55px","lineHeight":3.5,"fontSize":"55px","textDecoration":"line-through"}},"fontFamily":"source-serif-pro"} /-->

<!-- wp:separator {"className":"is-style-wide"} -->
<hr class="wp-block-separator has-alpha-channel-opacity is-style-wide"/>
<!-- /wp:separator -->

<!-- wp:search {"label":"Search","buttonText":"Search","buttonPosition":"no-button","style":{"typography":{"fontStyle":"italic","fontWeight":"600","textTransform":"capitalize","letterSpacing":"55px","lineHeight":3.5,"fontSize":"55px","textDecoration":"line-through"}},"fontFamily":"source-serif-pro"} /-->

```

</details>

Example theme.json snippet.

```json
	"styles": {
		"blocks": {
			"core/search": {
				"typography": {
					"textDecoration": "line-through"
					"letterSpacing": "12px",
					"fontSize": "2rem",
					"textTransform": "uppercase",
					"lineHeight": "1",
					"fontStyle": "italic",
					"fontWeight": "600"
				}
			}
		}
	},
```

## Screenshots or screencast <!-- if applicable -->


https://user-images.githubusercontent.com/6458278/186588035-43e0c91c-10c8-4bd5-90bf-5937c508a878.mp4

ramonjd committed:
Adding typography block supports to the search block
Ensuring that letter spacing is inherited.

aaronrobertshaw reviewed (commented):
Thanks for tackling typography on the search block  @ramonjd 👍 

In general, this works for me in the editor however I hit a couple of small issues on the frontend.

If we continue with the approach of applying the typography support styles to the root of the Search block, we'll need to also inherit `font-weight`, `font-style`, and `text-transform` on the button as you did for `letter-spacing`.

| Editor | Frontend |
|---|---|
| <img width="674" alt="Screen Shot 2022-08-23 at 4 43 19 pm" src="https://user-images.githubusercontent.com/60436221/186091060-3dbc66b6-618d-44ad-a2af-5b558c0818ba.png"> | <img width="679" alt="Screen Shot 2022-08-23 at 4 43 12 pm" src="https://user-images.githubusercontent.com/60436221/186091095-a3b185d2-ff1c-4f77-8a66-5116e075f6c9.png"> |
 
 Also, I know text-decoration is an extreme edge case in most situations however it feels to me like it's more of a problem when a user is typing a value into the search input and it's appearing with the `line-through` style.
 
 Additionally, the input with text decoration appears to be rendered differently across browsers. Firefox ignored the text decoration, Safari only rendered the line through for the user entered text, and Chrome rendered it for both the placeholder and entered text.
 
 The current directive is to supply consistent typographical design tools (albeit optionally displayed). Would we be better served here by skipping the serialization of the typography supports and applying them manually to the label and button elements? Then leveraging the feature level selectors so that global styles and theme.json apply their styles only to those same elements.
 
 What do you think?

ramonjd commented:
> skipping the serialization of the typography supports and applying them manually to the label and button elements

I knew it was too good to be true. 😄 Thanks for testing. I'll sort it.

ramonjd committed:
Opting into all typography supports for the search block (with the exception of text-decoration for the input element) so we can test and eliminate features we think aren't required.

ramonjd commented:
I have the typography styles applied to the label, input (with the exception of text-decoration) and button just so we can test things out. Safari and Firefox looking okay!

I think the input does require at least font-size, otherwise things look a bit uneven:

<img width="863" alt="Screen Shot 2022-08-24 at 3 06 57 pm" src="https://user-images.githubusercontent.com/6458278/186333593-a6cf0883-7894-46b1-9033-707eff8f8c27.png">

### TODO

- [x] Global styles and feature selectors

ramonjd commented on packages/block-library/src/search/index.php:
```diff
@@ -51,9 +51,10 @@ function render_block_core_search( $attributes ) {
 			$input_id,
-			$label_inner_html
+			$inline_styles['label'],
+			$label_inner_html,
```
Will fix trailing comma fix later.

ramonjd commented on packages/block-library/src/search/edit.js:
```diff
@@ -451,12 +463,13 @@ export default function SearchEdit( {
 			{ showLabel && (
 				<RichText
-					className="wp-block-search__label"
+					className={ `wp-block-search__label ${ typographyProps.className }` }
```
Will move to `classnames()`

ramonjd committed:
There should be a classname for font presets
Formatting

ramonjd committed:
Adding selectors for global styles.

ramonjd committed:
Adding typography classes in the back end

ramonjd commented on packages/block-library/src/search/index.php:
```diff
@@ -360,13 +378,118 @@ function styles_for_block_core_search( $attributes ) {
 	}
 
+	// Add typography styles.
+	$has_font_size = ! empty( $attributes['style']['typography']['fontSize'] );
```
Note to self: this is possibly something the style engine could optimize, since it will do all the `$has_` checks internally, and return the classnames. 

E.g., 

```php
$typography_styles = gutenberg_style_engine_get_styles( array( 'typography' => $$attributes['style']['typography'] ) );

$button_styles = $typography_styles['css'];
$input_styles = $typography_styles['css'];

if ( $show_label ) {
    $label_styles = $typography_styles['css'];
}
```

Doesn't take into account no textDecoration for the input. 

Maybe a follow up refactor anyway....

carolinan commented:
Styling the placeholder text and the input field text can make the block less accessible, for example by using a too low contrast. I would prefer if these options did not apply to the input field at all. If that is not an option, is there at least a specific color contrast warning?

aaronrobertshaw commented on packages/block-editor/src/hooks/use-typography-props.js:
```diff
@@ -0,0 +1,40 @@
+import { getInlineStyles } from './style';
+
+// This utility is intended to assist where the serialization of the typography
+// block support is being skipped for a block but the spacing related CSS
```
```suggestion
// block support is being skipped for a block but the typography related CSS
```

Missed updating this on the copy/paste.

aaronrobertshaw commented on packages/block-editor/src/hooks/use-typography-props.js:
```diff
@@ -0,0 +1,40 @@
+		? `has-${ kebabCase( attributes.fontFamily ) }-font-family`
+		: '';
+
+	const fontSizeClassName = !! attributes?.fontSize
+		? `has-${ kebabCase( attributes.fontSize ) }-font-size`
+		: '';
```
Could we use the [`getFontSizeClass`](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/font-sizes/utils.js#L60) util rather than duplicate the code generating the class here? I guess there isn't a matching util for the font family, though, so maybe staying consistent here is better.

ramonjd replied:
> Could we use the getFontSizeClass util

Definitely. Thanks for pointing that one out.

aaronrobertshaw commented on packages/block-editor/src/index.js:
```diff
@@ -6,6 +6,7 @@ export {
 	getBorderClassesAndStyles as __experimentalGetBorderClassesAndStyles,
 	useBorderProps as __experimentalUseBorderProps,
 	getColorClassesAndStyles as __experimentalGetColorClassesAndStyles,
+	getTypographyClassesAndStyles as __experimentalGetTypographyClassesAndStyles,
```
I know the other util functions were exported as experimental but do you think we could export this as stable? We'll be looking to stabilize all the block support APIs soon.

aaronrobertshaw commented on packages/block-library/src/search/index.php:
```diff
@@ -360,13 +378,118 @@ function styles_for_block_core_search( $attributes ) {
 		'input'   => ! empty( $input_styles ) ? sprintf( ' style="%s"', safecss_filter_attr( implode( ' ', $input_styles ) ) ) : '',
 		'button'  => ! empty( $button_styles ) ? sprintf( ' style="%s"', safecss_filter_attr( implode( ' ', $button_styles ) ) ) : '',
 		'wrapper' => ! empty( $wrapper_styles ) ? sprintf( ' style="%s"', safecss_filter_attr( implode( ' ', $wrapper_styles ) ) ) : '',
+		'label'   => ! empty( $label_styles ) ? sprintf( ' style="%s"', safecss_filter_attr( implode( ' ', $label_styles ) ) ) : '',
```
```suggestion
		'label'   => ! empty( $label_styles ) ? sprintf( ' style="%s"', esc_attr( safecss_filter_attr( implode( ' ', $label_styles ) ) ) ) : '',
```

There's an open PR that should merge soon that already addresses the other styles. It won't address the new `label` styles though.

aaronrobertshaw reviewed (commented):
@ramonjd Thanks for taking the leap to skip the serialization of typography styles for the Search block. 🙇 

I'm not sure my suggestion to skip serialization and use feature-level selectors works unless we skip applying all typography from the input. As this PR stands now, if you configure text-decoration via theme.json, it will still be applied to the input.

You can test this via the snippet below:

```json
			"core/search": {
				"typography": {
					"textDecoration": "line-through"
				}
			}
```

I agree with your [earlier comment](https://github.com/WordPress/gutenberg/pull/43499#issuecomment-1225200365) though about things looking odd and out-of-whack if we don't apply the non-decoration typography styles to the input. 

At this stage, I don't believe we can exclude text-decoration styles from being included in the theme.json/global styles generated stylesheets. So short of adding that, we could override it via the block's CSS. However, due to the specificity of the generated selector, we'd need to fight that if taking this approach.

I appreciate the next suggestion will be a pain, but maybe we need to tweak the approach again. Could we do something similar to the following?

- Continue to skip serialization of typography supports
- Remove use of feature-level selectors
- Apply the typography supports (excluding text-decoration) on the root wrapper
- Apply text decoration to the button and label if appropriate
- Add CSS override (which should only have to be now more specific than `.wp-block-search`)

What do you think?

I'd be happy to help on this or create an alternative PR for comparison if you think it would help.

---

> Styling the placeholder text and the input field text can make the block less accessible, for example by using a too low contrast

@carolinan, unless I'm misunderstanding something, this PR should only be applying typography styles to the input. The color block support styles remain unchanged.

That said, I appreciate that color isn't the only characteristic that might impact accessibility. How much of an issue is changing font size or weight? Particularly, if it matches the button as per Ramon's [earlier example](https://github.com/WordPress/gutenberg/pull/43499#issuecomment-1225200365)?

ramonjd committed:
Update packages/block-editor/src/hooks/use-typography-props.js

Co-authored-by: Aaron Robertshaw <60436221+aaronrobertshaw@users.noreply.github.com>

ramonjd committed:
Update packages/block-library/src/search/index.php

Co-authored-by: Aaron Robertshaw <60436221+aaronrobertshaw@users.noreply.github.com>

ramonjd commented:
> I appreciate the next suggestion will be a pain, but maybe we need to tweak the approach again. Could we do something similar to the following?

No problems at all. Thanks for retesting and the advice. I'll try to work things in the direction you mentioned.

ramonjd committed:
Apply text decoration only to label and button
Apply all typography styles except text decoration to block wrapper
Inherit block wrapper styles in CSS for each element
Use getFontSizeClass()
Remove __experimentalSelector
Make getTypographyClassesAndStyles() stable

ramonjd commented:
Failing tests seem related.
`waiting for selector "iframe[title="Site Preview"] >> nth=-1 >> control=enter-frame >> css=.widget-content >> role=searchbox[name="My Search"]"`
 👀

ramonjd committed:
The attributes were in the wrong order! Who would have thunk it?!

carolinan commented:
>  How much of an issue is changing font size or weight? Particularly, if it matches the button as per Ramon's [earlier example](https://github.com/WordPress/gutenberg/pull/43499#issuecomment-1225200365)?

I think it was the font weight that I got confused with color, as in the example, the light font weight (left image) is barely readable. Font size should not be an issue (disregarding cases where a user choose a font size that is too small in general)

aaronrobertshaw commented on packages/block-library/src/search/edit.js:
```diff
@@ -208,11 +210,14 @@ export default function SearchEdit( {
-		const textFieldStyles = isButtonPositionInside
-			? { borderRadius }
-			: borderProps.style;
+		const textFieldStyles = {
+			...( isButtonPositionInside
+				? { borderRadius }
+				: borderProps.style ),
+		};
```
After rolling back previous changes, I don't think we need this change anymore.

ramonjd replied:
Thanks for testing again @aaronrobertshaw, and for spotting the unnecessary code snippet here.

 It made me see something else: I'd forgotten to roll back the application of the font size and family preset classnames to the individual elements. I've applied them to the wrapper only now.

You shouldn't see any differences in testing.

If it looks okay on Monday, please feel free to merge this PR. 🙇 🙇 🙇 🙇

aaronrobertshaw reviewed (approved):
Props for the continued iterations on this while wrangling everything else you are doing for 6.1 @ramonjd 🙇 Nice work.

✅ Typography styles are appropriately applied for individual blocks in both editors and the front end
✅ Search block typography can be styled via theme.json or Global Styles (with small compromise)

As we've encountered previously, without using a feature-level selector, we can't prevent theme.json or global styles text decoration from being applied to the input. The downside of omitting styles for the input, resulting in its font size not matching the label and button, is more severe. I think the current approach is a fair compromise, particularly given how extremely niche any use of text-decoration would be for this block.

I also left an inline comment regarding a small unnecessary change we should remove before landing this.

ramonjd committed:
Ensure typography font size and family classnames appear on the wrapper, not on the individual elements.

aaronrobertshaw committed:
Clean up unnecessary destructuring

aaronrobertshaw committed:
Revert "Clean up unnecessary destructuring"

This reverts commit f778d690a7799b4c3cb207d12ca8271eba69569e.

aaronrobertshaw committed:
Revert "Ensure typography font size and family classnames appear on the wrapper, not on the individual elements."

This reverts commit 384ce7dadcafc3d44ff432257bab570fcd0498d9.

aaronrobertshaw committed:
Switch approach to target inner elements again

This commit reverts the approach to an earlier option tried so we can more reliably apply user-made typography selections without simple theme styles overriding the previous inherit styles.

This includes changes refining the previous approach so that text decoration isn't applied to the input by theme.json or global styles. It also fixes a bug around font family class generation in the render callback.

aaronrobertshaw committed:
Keep linter happy

aaronrobertshaw reviewed (commented):
Thanks for refining this PR further, @ramonjd . 👍 

While re-testing, I found some themes adding a simple `.wp-block-search__button` style would break the application of user-made typography selections due to the low specificity of the inherit rule.

I wrestled a bit with trying to find a neat way to overcome this. Including:
- Applying just styles like `font-size` and `font-family` to the button and label elements. The result is pretty inconsistent in terms of where styles are being applied as well as which styles should be.
- Adding custom classnames that we could hook onto to increase the specificity of the inherit styles. This worked for individual blocks but not the styles generated by theme.json or global styles.
- Ultimately, I circled back on one of your previous approaches where we apply all the styles to the inner elements and use feature-level selectors so the theme.json and global styles generated styles target the inner elements.
    - Without text-decoration being applied on the parent we can override the theme.json and global styles generated class and forcibly unset the text-decoration for the input.
    - Applying the styles on the inner elements helps mitigate the risk of themes such as TwentyTwentyTwo overriding user-made selections.
    - Via the feature-level selectors we can better control the specificity of the search block's styles via theme.json and global styles.
    
    Given the heavy changes I've made to the PR and going around in circles a bit. I think we need some fresh eyes to review this again.

tellthemachines commented on packages/block-editor/src/hooks/test/use-typography-props.js:
```diff
@@ -0,0 +1,28 @@
+			style: {
+				letterSpacing: '22px',
+				fontSize: '2rem',
+				fontFamily: 'tahini',
```
Not sure I get this. So we could have an element with a `has-tofu-font-family` classname and an actual font-family of `tahini`? Shouldn't the classname be generated from the style?

aaronrobertshaw replied:
The font-family class is generated from the `fontFamily` block attribute. That's done independently from whatever is in the style attribute. The inline styles, however, are generated solely from the style attribute. My assumption of @ramonjd's intent here was that it checked they didn't impact each other.

In terms of how the block supports themselves function, I don't think we can have a block with a named font family (i.e. `fontFamily` attribute set) as well as an inline style for one (`style.typography.fontFamily`).

tellthemachines replied:
Yeah, I'd expect that the presence of an inline style _would_ impact the classnames because they shouldn't both coexist if that makes sense? Though I get this function doesn't touch that logic, as documentation the test feels confusing. Not a blocker though 😅

aaronrobertshaw replied:
I agree on both counts. 

You could also argue that the presence of an inline style for our block supports does impact the classname as when it is set the top-level `fontFamily` block attribute gets cleared, effectively clearing the classname.

I'll replace the `fontFamily` inline style with something else to avoid further confusion.

aaronrobertshaw replied:
Tweaked the test via https://github.com/WordPress/gutenberg/pull/43499/commits/d0aa4a3cb6ebee75c3ec83cf2320f9e68bd2048e

aaronrobertshaw committed:
Reapply clean up and force text-decoration unset

tellthemachines commented on packages/block-editor/src/index.js:
```diff
@@ -6,6 +6,7 @@ export {
 	getBorderClassesAndStyles as __experimentalGetBorderClassesAndStyles,
 	useBorderProps as __experimentalUseBorderProps,
 	getColorClassesAndStyles as __experimentalGetColorClassesAndStyles,
+	getTypographyClassesAndStyles,
```
Should we be marking this experimental?

aaronrobertshaw replied:
@ramonjd initially had this as experimental until I suggested it be stabilized as we'll be stabilizing all experimental block support APIs post-6.1. This would just be another experimental API to process. Given it follows the same pattern as the other block supports, e.g. color and border, and they're due for stabilization. I figure we can save double handling this one.

We can change this if you have strong concerns.

tellthemachines replied:
That makes sense to me! If color/border are similar and we're not planning on making further changes before stabilising, I wouldn't be opposed to doing it pre-6.1 either. Or were we aiming to stabilise all block supports together? It might take a while to get layout up to spec 😬

aaronrobertshaw replied:
To my knowledge, there are no plans for further changes to these util functions. I was under the impression that we'd stabilize each block support in full, e.g. all the typography support APIs, then color, and so on. Layout support has always been a special case, so I don't think there is a rush to stabilize it before it's ready.

Regarding stabilizing the others pre-6.1, I don't think that's on the cards. My focus has been on adopting the block supports more consistently within Gutenberg and then seeing if there is anything we should alter before stabilizating them.

That said, if there's a window in which we could actually get it all done, I wouldn't be against it.

tellthemachines reviewed (commented):
Looks pretty close! It's working well in testing, except for a small detail: with big line heights the button icon is slightly off-center:
<img width="721" alt="Screen Shot 2022-09-12 at 4 30 21 pm" src="https://user-images.githubusercontent.com/8096000/189587685-a1259bdf-f91e-411c-bf1e-b0b1453c3a31.png">
Adding `vertical-align: text-bottom;` to `.wp-block-search__button svg` seems to fix it. 

Apart from that, just a couple questions below.

aaronrobertshaw commented:
> with big line heights the button icon is slightly off-center

Good catch @tellthemachines, I'll tweak the styles accordingly. Thanks.

aaronrobertshaw committed:
Avoid confusion over font-family being in classes and inline styles

aaronrobertshaw committed:
Fix vertical alignment of icon only button

This centres the icon for the icon only button on the frontend when it has a tall line-height.

tellthemachines reviewed (approved):
Thanks, LGTM!
